# Data Manipulation

### Importing the data into a Data Frame

In [1]:
import pandas as pd
import numpy as np

In [2]:
df=pd.read_csv('data.csv')
df.head()

Customer_id  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0            1           1.1          93.994          -36.4      4.857   
1            2           1.1          93.994          -36.4      4.857   
2            3           1.1          93.994          -36.4      4.857   
3            4           1.1          93.994          -36.4      4.857   
4            5           1.1          93.994          -36.4      4.857   

   nr.employed  age          job  marital            education  ...  \
0       5191.0   56     services  married          high.school  ...   
1       5191.0   45     services  married             basic.9y  ...   
2       5191.0   59       admin.  married  professional.course  ...   
3       5191.0   41  blue-collar  married              unknown  ...   
4       5191.0   24   technician   single  professional.course  ...   

  Region_Name    contact month day_of_week duration  campaign pdays previous  \
0       South  telephone   may         mon      307         1   999        0   
1       South  telephone   may         mon      198         1   999        0   
2        West  telephone   may         mon      139         1   999        0   
3       South  telephone   may         mon      217         1   999        0   
4       South  telephone   may         mon      380         1   999        0   

      poutcome Customer_Response  
0  nonexistent                no  
1  nonexistent                no  
2  nonexistent                no  
3  nonexistent                no  
4  nonexistent                no  

[5 rows x 29 columns]

In [3]:
df.columns

Index(['Customer_id', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx',
       'euribor3m', 'nr.employed', 'age', 'job', 'marital', 'education',
       'default', 'housing', 'loan', 'State_Code', 'City_Code', 'Postal Code',
       'City_Name', 'State_Name', 'Region_Code', 'Region_Name', 'contact',
       'month', 'day_of_week', 'duration', 'campaign', 'pdays', 'previous',
       'poutcome', 'Customer_Response'],
      dtype='object')

### Drop all the rows with missing values
Here we have missing values labelled unknown

In [4]:
df = df[~df.isin(['unknown']).any(axis=1)]
df.shape

(27494, 29)

In [5]:
#df.to_csv('dropped.csv', index = False)

For further analysis keep the data handy into a separate file(csv)

### `Analysis of percentage turnout of marketing campaign`

In [6]:
pd.DataFrame({'Actual Counts': df['Customer_Response'].value_counts(),
              'Conversion Percentage': (df['Customer_Response'].value_counts(normalize=True))*100})

Actual Counts  Conversion Percentage
no           23995              87.273587
yes           3499              12.726413

- There is a 12.7% conversion rate which is good in industry comparison
- But further analysis is required to decide if all campaigns are equally contributing

In [7]:
pd.DataFrame({'Actual Counts': df.groupby('Customer_Response')['campaign'].value_counts(),
              'Percentage': (df.groupby('Customer_Response')['campaign'].value_counts(normalize=True))*100})

Actual Counts  Percentage
Customer_Response campaign                           
no                1                 10191   42.471348
                  2                  6193   25.809544
                  3                  3063   12.765159
                  4                  1572    6.551365
                  5                   949    3.954991
                  6                   574    2.392165
                  7                   378    1.575328
                  8                   244    1.016879
                  9                   170    0.708481
                  10                  133    0.554282
                  11                  108    0.450094
                  12                   82    0.341738
                  13                   52    0.216712
                  14                   41    0.170869
                  17                   32    0.133361
                  16                   28    0.116691
                  15                   24    0.100021
                  18                   22    0.091686
                  20                   21    0.087518
                  19                   17    0.070848
                  21                   14    0.058345
                  23                   11    0.045843
                  22                   10    0.041675
                  24                   10    0.041675
                  29                    7    0.029173
                  27                    6    0.025005
                  28                    6    0.025005
                  30                    6    0.025005
                  25                    4    0.016670
                  26                    4    0.016670
                  31                    4    0.016670
                  32                    4    0.016670
                  33                    3    0.012503
                  35                    3    0.012503
                  34                    2    0.008335
                  40                    2    0.008335
                  37                    1    0.004168
                  39                    1    0.004168
                  41                    1    0.004168
                  42                    1    0.004168
                  43                    1    0.004168
yes               1                  1740   49.728494
                  2                   924   26.407545
                  3                   429   12.260646
                  4                   188    5.372964
                  5                    93    2.657902
                  6                    49    1.400400
                  7                    26    0.743069
                  8                    13    0.371535
                  9                    11    0.314376
                  10                    9    0.257216
                  11                    9    0.257216
                  17                    4    0.114318
                  12                    2    0.057159
                  14                    1    0.028580
                  23                    1    0.028580

- More no. of calls have not given good results
- It is suggested to keep number of contacts limited to 2
    - As more no. of calls may create a negeative perspective in the customers
    - Calls could also be reported as spam

### `Right mode to contact the customers (telephone or mobile) `

In [8]:
pd.DataFrame({'Actual Counts': df.groupby('contact')['Customer_Response'].value_counts(),
              'Percentage': (df.groupby('contact')['Customer_Response'].value_counts(normalize=True))*100})

Actual Counts  Percentage
contact   Customer_Response                           
cellular  no                         15461   83.876743
          yes                         2972   16.123257
telephone no                          8534   94.183865
          yes                          527    5.816135

- A higher conversion rate is observed in the cellular mode of contact.
- Greater emphasis is recommended on the cellular channel.
- Telephone might be a redundant channel; however, this hypothesis needs to be confirmed with additional context

### `Analysis on attempts made to turn a person into successful depositor`

In [9]:
print('Analysis of Campaign Response based on previous response')
p =pd.crosstab(df['previous'], df['Customer_Response'], normalize = 'index')
p['Count']=df['previous'].value_counts()
p

Analysis of Campaign Response based on previous response


Customer_Response        no       yes  Count
previous                                    
0                  0.900116  0.099884  23307
1                  0.775637  0.224363   3374
2                  0.526224  0.473776    572
3                  0.390533  0.609467    169
4                  0.461538  0.538462     52
5                  0.333333  0.666667     15
6                  0.500000  0.500000      4
7                  1.000000  0.000000      1

### There are a few patterns of correlation
- After 2 attempts apart from 7 there has always more positive responses
- It seems that attempting more times is giving better conversion
- The data is not equally distributed among all the attempts
- Hence nothing can be constructively concluded from this exploration

In [10]:
# Total number of campaigns
total_campaigns = df['campaign'].count()

# Number of successful deposits
successful_deposits = len(df[df['Customer_Response'] == 'yes'])

# Success rate
success_rate = (successful_deposits / total_campaigns) * 100

# Percentage of clients previously contacted
previously_contacted_percentage = (len(df[(df['previous'] != 0) & (df['Customer_Response']=='yes')]) / len(df)) * 100
previously_not_contacted_percentage = (len(df[(df['previous'] == 0) & (df['Customer_Response']=='yes')]) / len(df)) * 100

print(f"Total Campaigns: {total_campaigns}")
print(f"Number of Successful Deposits: {successful_deposits}")
print(f"Success Rate: {success_rate:.2f}%")
print(f"Percentage of Clients Previously Contacted: {previously_contacted_percentage:.2f}%")
print(f"Percentage of Clients Previously Not Contacted: {previously_not_contacted_percentage:.2f}%")

Total Campaigns: 27494
Number of Successful Deposits: 3499
Success Rate: 12.73%
Percentage of Clients Previously Contacted: 4.26%
Percentage of Clients Previously Not Contacted: 8.47%


#### Though it appears that conversion is not high multiple contacts is still advisable
- The people in previously contacted category are less in number, hence better ROI (Return On Investment)
- It gives a sense of connection for customers when call back is done to make sure they are valued

### `Personal data analysis on marital status, existing loans, education, profession etc. and its impact on the campaign’s success. `

In [11]:
def crosstab(df, col1, col2):
    data = pd.crosstab(df[col1], df[col2], normalize = 'index')
    data['Count'] = df[col1].value_counts()
    return data

In [12]:
crosstab(df, 'marital', 'Customer_Response')

Customer_Response        no       yes  Count
marital                                     
divorced           0.884266  0.115734   3197
married            0.882476  0.117524  15801
single             0.850282  0.149718   8496

#### Single people have better conversion rates, hence focus can be put on this aspect

In [13]:
crosstab(df, 'loan', 'Customer_Response')

Customer_Response        no       yes  Count
loan                                        
no                 0.871848  0.128152  23199
yes                0.877532  0.122468   4295

#### Loan status seems to be neutral in terms of target customers
- There is a slight favour observed in conversion of customers with no loan
- Or it could also be interpreted as existing loan might not be a good contributor as per primary analysis

In [14]:
crosstab(df, 'education', 'Customer_Response')

Customer_Response          no       yes  Count
education                                     
basic.4y             0.863026  0.136974   2161
basic.6y             0.900398  0.099602   1255
basic.9y             0.912048  0.087952   3843
high.school          0.876189  0.123811   6938
illiterate           0.888889  0.111111      9
professional.course  0.874904  0.125096   3901
university.degree    0.851710  0.148290   9387

- University graduates are the best target customers with the highest conversion rate.
- This is closely followed by customers with a basic four-year education.
- Tailoring marketing to these segments is advisable.

In [15]:
crosstab(df, 'job', 'Customer_Response')

Customer_Response        no       yes  Count
job                                         
admin.             0.858065  0.141935   7898
blue-collar        0.920430  0.079570   5115
entrepreneur       0.904811  0.095189    977
housemaid          0.878247  0.121753    616
management         0.878352  0.121648   2088
retired            0.703398  0.296602   1089
self-employed      0.882773  0.117227    981
services           0.906813  0.093187   2554
student            0.670391  0.329609    537
technician         0.883632  0.116368   4967
unemployed         0.824405  0.175595    672

- Students seem to have the highest conversion rate.
- Focus on students may be advisable.
- If the business perceives students as non-reliable, it is better to expand campaigning limits to suit all kinds of customers.
- Retired customers can also be good targets as they also have a decent conversion rate.

## `Socio-economic analysis of the customers`

1) `emp.var.rate: employment variation rate`

In [16]:
crosstab(df, 'emp.var.rate', 'Customer_Response')

Customer_Response        no       yes  Count
emp.var.rate                                
-3.4               0.572110  0.427890    839
-3.0               0.496241  0.503759    133
-2.9               0.633846  0.366154   1300
-1.8               0.827700  0.172300   6686
-1.7               0.481600  0.518400    625
-1.1               0.519841  0.480159    504
-0.2               0.888889  0.111111      9
-0.1               0.934968  0.065032   2814
 1.1               0.968890  0.031110   4468
 1.4               0.945038  0.054962  10116

### There is a good mix of target customers in terms of employment variation rate
- Impact of this on buisiness might unpredictable as there is no definitive pattern
- However it seems like the campaign is appealing better to customers with negative emp.var.rate

#### `cons.price.idx: consumer price index`

In [17]:
crosstab(df, 'cons.price.idx', 'Customer_Response')

Customer_Response        no       yes  Count
cons.price.idx                              
92.201             0.655684  0.344316    607
92.379             0.603960  0.396040    202
92.431             0.594752  0.405248    343
92.469             0.604167  0.395833    144
92.649             0.523810  0.476190    294
92.713             0.496241  0.503759    133
92.756             0.888889  0.111111      9
92.843             0.538793  0.461207    232
92.893             0.902994  0.097006   4175
92.963             0.617486  0.382514    549
93.075             0.801124  0.198876   1780
93.200             0.947731  0.052269   2755
93.369             0.427136  0.572864    199
93.444             0.948643  0.051357   3427
93.749             0.432836  0.567164    134
93.798             0.338983  0.661017     59
93.876             0.421687  0.578313    166
93.918             0.935530  0.064470   4188
93.994             0.968890  0.031110   4468
94.027             0.483333  0.516667    180
94.055             0.539267  0.460733    191
94.199             0.514644  0.485356    239
94.215             0.437008  0.562992    254
94.465             0.956018  0.043982   2501
94.601             0.515924  0.484076    157
94.767             0.537037  0.462963    108

### There is no definitive pattern yet:
- As CPI moves away from 100 there is higher rate of conversion
- This confirms our previous hypothesis that customers with negative emp.var.rate are better target customers

#### `cons.conf.idx: consumer confidence index`

In [18]:
crosstab(df, 'cons.conf.idx', 'Customer_Response')

Customer_Response        no       yes  Count
cons.conf.idx                               
-50.8              0.537037  0.462963    108
-50.0              0.538793  0.461207    232
-49.5              0.515924  0.484076    157
-47.1              0.801124  0.198876   1780
-46.2              0.902994  0.097006   4175
-45.9              0.888889  0.111111      9
-42.7              0.935530  0.064470   4188
-42.0              0.947731  0.052269   2755
-41.8              0.956018  0.043982   2501
-40.8              0.617486  0.382514    549
-40.4              0.338983  0.661017     59
-40.3              0.437008  0.562992    254
-40.0              0.421687  0.578313    166
-39.8              0.539267  0.460733    191
-38.3              0.483333  0.516667    180
-37.5              0.514644  0.485356    239
-36.4              0.968890  0.031110   4468
-36.1              0.948643  0.051357   3427
-34.8              0.427136  0.572864    199
-34.6              0.432836  0.567164    134
-33.6              0.604167  0.395833    144
-33.0              0.496241  0.503759    133
-31.4              0.655684  0.344316    607
-30.1              0.523810  0.476190    294
-29.8              0.603960  0.396040    202
-26.9              0.594752  0.405248    343

#### There is no definitive pattern observed
- This data might not give useful insights for campaign
- The whole data consisted of a negative cons.conf.idx
- Low consumer confidence may lead to reduced spending, potentially affecting economic activity.
- Hence we should try to include more customers of positive confidence level

#### `euribor 3m: euribor 3 month rate`

In [19]:
df['euribor3m'].describe()

count    27494.000000
mean         3.462257
std          1.776449
min          0.634000
25%          1.313000
50%          4.856000
75%          4.961000
max          5.045000
Name: euribor3m, dtype: float64

#### High euribor is good for deposit schemes which most of our customers are offered
- The median and 75th percentile are really close indicating the data is right skewed in distribution
- The mean is slightly lacking to the median but is still significant

#### `nr.employed: number of employees `

In [20]:
crosstab(df, 'nr.employed', 'Customer_Response')

Customer_Response        no       yes  Count
nr.employed                                 
4963.6             0.519841  0.480159    504
4991.6             0.481600  0.518400    625
5008.7             0.426854  0.573146    499
5017.5             0.572110  0.427890    839
5023.5             0.496241  0.503759    133
5076.2             0.633846  0.366154   1300
5099.1             0.860029  0.139971   6187
5176.3             0.888889  0.111111      9
5191.0             0.968890  0.031110   4468
5195.8             0.934968  0.065032   2814
5228.1             0.945038  0.054962  10116

- The count of 10116 is for customers with nr.emplyment as 5228 which is very high
- Still there is not a high conversion rate
- Hence we can make the target group more diverse to improve the conversion rates

## `Demographic analysis of the marketing campaign using the master files`

In [22]:
city_master = pd.read_csv('City_Master.csv')
state_master = pd.read_csv('State_Master.csv')
region_code_master = pd.read_csv('Region_code_master.csv')
customer_campaign_details = pd.read_csv('Customer_campaign_details_p1.csv')

In [23]:
master_data = pd.merge(state_master, region_code_master, on='Region_Code')
master_data = pd.merge(master_data, city_master, on='State_Code')
master_data.head()

State_Code State_Name  Region_Code Region_Name City_Code      City_Name
0         S1   Kentucky            3       South        C1      Henderson
1         S1   Kentucky            3       South      C103       Richmond
2         S1   Kentucky            3       South      C160       Florence
3         S1   Kentucky            3       South      C209         Murray
4         S1   Kentucky            3       South      C257  Bowling Green

In [24]:
(df['State_Name'].value_counts()).head(10)

California        5526
New York          3111
Texas             2686
Pennsylvania      1644
Washington        1379
Illinois          1331
Ohio              1296
Florida           1011
North Carolina     701
Michigan           695
Name: State_Name, dtype: int64

#### The customer base has strong domination in California, New York and Texas
- Further analysis required on how campaigns are performing in these particular areas
- We can focus on spreading the target region further more in terms of spread

In [31]:
df['Region_Name'].value_counts()

West       8862
East       7830
Central    6342
South      4460
Name: Region_Name, dtype: int64

#### Strong focus on West and East Regions
- Data has better spread in terms of region than state

In [26]:
(df['City_Name'].value_counts()).head(10)

New York City    2535
Los Angeles      2083
Philadelphia     1518
San Francisco    1420
Seattle          1159
Houston          1024
Chicago           854
Columbus          626
Springfield       444
Dallas            443
Name: City_Name, dtype: int64

#### New York City and Los Angeles are the top cities of focus

## `Other Primary Analysis`

In [27]:
df['age'].describe()

count    27494.000000
mean        39.060413
std         10.343971
min         17.000000
25%         31.000000
50%         37.000000
75%         45.750000
max         95.000000
Name: age, dtype: float64

- The age range of customers is 17 to 98, indicating a diverse range of customers.
- The median age is 38, representing the working class, which makes them a good target for a finance-related campaign.

In [28]:
df['marital'].value_counts(normalize=True)

married     0.574707
single      0.309013
divorced    0.116280
Name: marital, dtype: float64

#### Majority of the people are married but from earlier hypothesis keeping more single people in target would be benificial.
- Hence re-validating the target segment would be better

In [29]:
df['education'].value_counts(normalize=True)

university.degree      0.341420
high.school            0.252346
professional.course    0.141886
basic.9y               0.139776
basic.4y               0.078599
basic.6y               0.045646
illiterate             0.000327
Name: education, dtype: float64

- People with a university degree are good target customers, aligning with our earlier observation.
- However, since illiterate people also had a high conversion rate, there could be more of them in the target market.

In [30]:
df['job'].value_counts(normalize=True)

admin.           0.287263
blue-collar      0.186041
technician       0.180658
services         0.092893
management       0.075944
retired          0.039609
self-employed    0.035681
entrepreneur     0.035535
unemployed       0.024442
housemaid        0.022405
student          0.019532
Name: job, dtype: float64

#### People with admin profiles and blue-collar jobs are most of the targeted customers

#### `Primary data manipulation shows of the following`
- There are many campaigns with no conversion at all
- The customer connects are not equally distributed among the campaigns
- Cellular mode is best to connect with customers
- Attempting again in a few days seems favouring conversion
- Previous contacts are less in no yet high in conversion
- The following customers are in highly converted customers
    - Single in marital status
    - No loan existing
    - Educated with basic 4y
    - Students in terms of education